src/util/LimitedAllocator.h

```diff
@@ -0,0 +1,164 @@
+ * same time
+ *
+ * The actual allocation is done by a std::allocator<T> which is wrapped by
+ * checking the limit
+ *
+ * Memory allocated by copies of an Allocator will also count towards the limit
+ * To use it, construct a first allocator by calling
+ * LimitedAllocator{makeAllocationState(limit)} and then pass copies of this
+ * allocator to your containers. You can also Create Allocators for different
+ * types respecting the same total memory pool size, e.g.
+ *
+ * auto state = makeAllocationState(limitInBytes);
+ * auto allocInt = LimitedAllocator<int>{state};
+ * auto limitedIntVec = std::vector<int, LimitedAllocator<int>>{allocInt};
+ * auto allocString = LimitedAllocator<string>{state};
+ * auto limitedStringVec = std::vector<string,
+ * LimitedAllocator<string>>{allocString};
+ *
+ * // now the total amount of memory allocated by limitedIntVec and
+ * limitedStringVec may never exceed limitInBytes
+ *
+ * @tparam T the type of Elements that this allocator allocates
+ */
+template <typename T>
+class LimitedAllocator {
+ public:
+  using value_type = T;
+
+  // These special type aliases are necessary for using this allocator with
+  // the STL. We always want it to be propagated to the target of copy and move
+  // operations, s.T. the copy also counts towards the Limit.
+  using propagate_on_container_copy_assignment = std::true_type;
+  using propagate_on_container_move_assignment = std::true_type;
+  using propagate_on_container_swap = std::true_type;
+
+ private:
+  detail::AllocationState state_;  // shared number of free bytes
+  std::allocator<T> alloc_;
+
+ public:
+  /// obtain an AllocationState by calls to makeAllocationState()
+  explicit LimitedAllocator(detail::AllocationState s) : state_(std::move(s)) {}
+  LimitedAllocator() = delete;
+
+  // TODO<C++20> : the exact signature of allocate changes
+  // the required interface of an allocator
+  T* allocate(std::size_t n) {
+    // first check if we have enough space and reserve it (otherwise throw)
+    state_.ptr()->wlock()->allocate(n * sizeof(T));
+    // the actual allocation
+    return alloc_.allocate(n);
+  }
+
+  // the required interface of an allocator
+  void deallocate(T* p, std::size_t n) {
+    // free the memory
+    alloc_.deallocate(p, n);
+    // say that we have indeed freed the memory, and others may reserve it
+    // again.
+    state_.ptr()->wlock()->deallocate(n * sizeof(T));
+  }
+
+  /// Return the number of bytes, that this allocator and all of its copies
+  /// currently have available
+  [[nodiscard]] size_t numFreeBytes() const {
+    // casting is ok, because the actual numFreeBytes call
+    // is const, and everything else is locking
+    return const_cast<LimitedAllocator*>(this)
+        ->state_.ptr()
+        ->wlock()
+        ->numFreeBytes();
+  }
+
+  // comparison of allocators for (in)-equality are necessary for the STL
+  // allocator interface allocators that point to the same state are equal;
```

Comment:
// The STL needs two allocators to be equal if and only they refer to the same memory pool. For us, they are hence equal if they use the same AllocationMemoryLeft object.